### ISSUE TITLE
Freeze Weights

### ISSUE THREAD

**Issue by ThorrenKirschbaum** [atomistic-machine-learning/schnetpack#511]
Hi,

I am following the PyTorch Lightning + SchnetPack code from the tutorial to train a PaiNN on energies and forces. Is it possible to freeze the first few layers of the model during the training? 

Many thanks in advance and best
Thorren

**Comment by Stefaanhess** [comment]
Hi @ThorrenKirschbaum 

yes, this should be possible to do. I have not tested it myself, but you could try the `freeze()` function of pytorch-lightning. They have a very basic [tutorial](https://lightning.ai/docs/pytorch/stable/advanced/transfer_learning.html) for this.

Best,
Stefaan

**Comment by ThorrenKirschbaum** [comment]
Thank you Stefaan, that seems to work. I will leave my code here for others to use / comment on.

class PaiNNTransferLearning(nn.Module): 
    def __init__(
        self,
        n_atom_basis: int,
        n_interactions: int,
        radial_basis: nn.Module,
        cutoff_fn: Optional[Callable] = None,
        activation: Optional[Callable] = F.silu,
        max_z: int = 100,
        shared_interactions: bool = False,
        shared_filters: bool = False,
        epsilon: float = 1e-8,
    ):
        """
        Args:
            n_atom_basis: number of features to describe atomic environments.
                This determines the size of each embedding vector; i.e. embeddings_dim.
            n_interactions: number of interaction blocks.
            radial_basis: layer for expanding interatomic distances in a basis set
            cutoff_fn: cutoff function
            activation: activation function
            shared_interactions: if True, share the weights across
                interaction blocks.
            shared_interactions: if True, share the weights across
                filter-generating networks.
            epsilon: stability constant added in norm to prevent numerical instabilities
        """
        super(PaiNNTransferLearning, self).__init__()

        backbone = torch.load("../best_inference_model_spce")

        self.n_atom_basis = n_atom_basis
        self.n_interactions = n_interactions
        self.cutoff_fn = cutoff_fn
        self.cutoff = cutoff_fn.cutoff
        self.radial_basis = radial_basis

        self.embedding = nn.Embedding(max_z, n_atom_basis, padding_idx=0)

        self.share_filters = shared_filters

        if shared_filters:
            self.filter_net = snn.Dense(
                self.radial_basis.n_rbf, 3 * n_atom_basis, activation=None
            )
        else:
            self.filter_net = snn.Dense(
                self.radial_basis.n_rbf,
                self.n_interactions * n_atom_basis * 3,
                activation=None,
            )
        self.interactions = snn.replicate_module(
            lambda: PaiNNInteraction(
                n_atom_basis=self.n_atom_basis, activation=activation
            ),
            self.n_interactions,
            shared_interactions,
        )
        self.mixing = snn.replicate_module(
            lambda: PaiNNMixing(
                n_atom_basis=self.n_atom_basis, activation=activation, epsilon=epsilon
            ),
            self.n_interactions,
            shared_interactions,
        )

    def forward(self, inputs: Dict[str, torch.Tensor]):
        """
        Compute atomic representations/embeddings.

        Args:
            inputs (dict of torch.Tensor): SchNetPack dictionary of input tensors.

        Returns:
            torch.Tensor: atom-wise representation.
            list of torch.Tensor: intermediate atom-wise representations, if
            return_intermediate=True was used.
        """
        # get tensors from input dictionary
        atomic_numbers = inputs[properties.Z]
        r_ij = inputs[properties.Rij]
        idx_i = inputs[properties.idx_i]
        idx_j = inputs[properties.idx_j]
        n_atoms = atomic_numbers.shape[0]

        # compute atom and pair features
        d_ij = torch.norm(r_ij, dim=1, keepdim=True)
        dir_ij = r_ij / d_ij
        phi_ij = self.radial_basis(d_ij)
        fcut = self.cutoff_fn(d_ij)

        filters = self.filter_net(phi_ij) * fcut[..., None]
        if self.share_filters:
            filter_list = [filters] * self.n_interactions
        else:
            filter_list = torch.split(filters, 3 * self.n_atom_basis, dim=-1)

        q = self.embedding(atomic_numbers)[:, None]
        qs = q.shape
        mu = torch.zeros((qs[0], 3, qs[2]), device=q.device)

        # here we freeze the network weights except for the last layer
        for i, (interaction, mixing) in enumerate(zip(self.interactions, self.mixing)):
            if i < len(self.interactions)-1:
                with torch.no_grad():
                    q, mu = interaction(q, mu, filter_list[i], dir_ij, idx_i, idx_j, n_atoms)
                    q, mu = mixing(q, mu)
            else:
                q, mu = interaction(q, mu, filter_list[i], dir_ij, idx_i, idx_j, n_atoms)
                q, mu = mixing(q, mu)

        q = q.squeeze(1)

        inputs["scalar_representation"] = q
        inputs["vector_representation"] = mu
        return inputs
